File: apps/mobile/modules/goproceed-vault/ios/GoProceedVaultModule.swift
import ExpoModulesCore

// Its own queue: an upload waits up to 185 s and must not hold Expo's shared
// AsyncFunction queue (SecureStore, camera) or delay cancel(). Concurrent because
// NativeVault serialises journal access itself and releases it during transfers.
private let vaultQueue = DispatchQueue(label: "com.lightholdlabs.goproceed.vault", qos: .userInitiated, attributes: .concurrent)

public class GoProceedVaultModule: Module {
  private let vault = NativeVault()
  public func definition() -> ModuleDefinition {
    Name("GoProceedVault")
    AsyncFunction("call") { (operation: String, payload: String) -> String in
      try self.vault.call(operation, payload)
    }.runOnQueue(vaultQueue)
    Function("cancel") { (quarantine: Bool) in self.vault.cancel(quarantine) }
    OnDestroy { self.vault.cancel(true) }
  }
}

File: apps/mobile/modules/goproceed-vault/android/src/main/java/expo/modules/goproceedvault/GoProceedVaultModule.kt
package expo.modules.goproceedvault

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob

// Its own scope: an upload can take minutes and must not hold Expo's shared
// AsyncFunction thread (other modules, SecureStore) or delay cancel(). Calls may
// run concurrently; NativeVault serialises journal access itself and releases it
// during a transfer.
private val vaultScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

class GoProceedVaultModule : Module() {
  private val vault by lazy { NativeVault.shared(requireNotNull(appContext.reactContext)) }
  override fun definition() = ModuleDefinition {
    Name("GoProceedVault")
    AsyncFunction("call") { operation: String, payload: String -> vault.call(operation, payload) }.runOnQueue(vaultScope)
    // Lock-only and synchronous: it must never wait behind a running call.
    Function("cancel") { quarantine: Boolean -> vault.cancel(quarantine) }
    OnDestroy { vault.cancel(true) }
  }
}
